tests/test_cargo_init.rs

```diff
@@ -0,0 +1,297 @@
+                                             env::consts::EXE_SUFFIX)),
+                existing_file());
+});
+
+fn bin_already_exists(explicit: bool, rellocation: &str) {
+    let path = paths::root().join("foo");
+    fs::create_dir(&path).ok();
+    fs::create_dir(&path.join("src")).ok();
+    
+    let sourcefile_path = path.join(rellocation);
+    
+    File::create(&sourcefile_path).unwrap().write_all(br#"
+        fn main() {
+            println!("Hello, world 2!");
+        }
+    "#).unwrap();
+    
+    if explicit {
+        assert_that(cargo_process("init").arg("--bin").arg("--vcs").arg("none")
+                                        .env("USER", "foo").cwd(&path),
+                    execs().with_status(0));
+    } else {
+        assert_that(cargo_process("init").arg("--vcs").arg("none")
+                                        .env("USER", "foo").cwd(&path),
+                    execs().with_status(0));
+    }
+
+    assert_that(&paths::root().join("foo/Cargo.toml"), existing_file());
+    assert_that(&paths::root().join("foo/src/lib.rs"), is_not(existing_file()));
+    
+    // Check that our file is not overwritten
+    let mut contents = String::new();
+    File::open(&sourcefile_path).unwrap().read_to_string(&mut contents).unwrap();
+    assert!(contents.contains(r#"Hello, world 2!"#));
+
+    assert_that(cargo_process("build").cwd(&paths::root().join("foo")),
+                execs().with_status(0));
+    assert_that(&paths::root().join(&format!("foo/target/debug/foo{}",
+                                             env::consts::EXE_SUFFIX)),
+                existing_file());
+}
+
+test!(bin_already_exists_explicit {
+    bin_already_exists(true, "src/main.rs")
+});
+
+test!(bin_already_exists_implicit {
+    bin_already_exists(false, "src/main.rs")
+});
+
+test!(bin_already_exists_explicit_nosrc {
+    bin_already_exists(true, "main.rs")
+});
+
+test!(bin_already_exists_implicit_nosrc {
+    bin_already_exists(false, "main.rs")
+});
+
+test!(bin_already_exists_implicit_namenosrc {
+    bin_already_exists(false, "foo.rs")
+});
+
+test!(bin_already_exists_implicit_namesrc {
+    bin_already_exists(false, "src/foo.rs")
+});
+
+test!(confused_by_multiple_lib_files {
+    let path = paths::root().join("foo");
+    fs::create_dir(&path).ok();
+    fs::create_dir(&path.join("src")).ok();
+    
+    let sourcefile_path1 = path.join("src/lib.rs");
+    
+    File::create(&sourcefile_path1).unwrap().write_all(br#"
+        fn qqq () {
+            println!("Hello, world 2!");
+        }
+    "#).unwrap();
+    
+    let sourcefile_path2 = path.join("lib.rs");
+    
+    File::create(&sourcefile_path2).unwrap().write_all(br#"
+        fn qqq () {
+            println!("Hello, world 3!");
+        }
+    "#).unwrap();
+    
+    assert_that(cargo_process("init").arg("--vcs").arg("none")
+                                    .env("USER", "foo").cwd(&path),
+                execs().with_status(101));
```

Comment:
1. Is it appropriate to deduplicate the error message between the test and the actual code? (e.g. reference to a constant from the test)?
2. Is it a good idea to test not the entire error message, but just a substring?